add youtube/video downloads and viewer

PThorpe92 opened the pull request:
This PR adds functionality to download videos from most popular video websites, so they can be viewed by students in the open content page/viewer.

![image](https://github.com/user-attachments/assets/e3b39411-f88f-447d-b9a4-e14ec52e5556)

The Provider Middleware is now equipped with three additional jobs, which it subscribes to via the NATS message queue.

1. Add videos:
   - This job accepts an array of video URLs input by an administrator, and attempts a download one at a time.
   
![image](https://github.com/user-attachments/assets/559787b2-abc3-49b1-ace6-e19b7c1d0f00)
![image](https://github.com/user-attachments/assets/03b5bcb2-929e-4110-9742-b95833e18ab5)

2. Retry failed downloads
  - This job is run automatically every 3 hours, and attempts to re-download any videos that have failed to download in the past. 
  - Each time a download is attempted, an `attempt` is added for that video, and after 5 attempts, the video will no longer retry, and the administrator will be prompted to remove the video, and the relevant error message will be displayed. 
 
3. Retry single video download
  - An administrator can select to attempt to retry the download manually, although the previous rule of 5 attempts maximum still applies. There is an endpoint at `/api/videos/{id}/retry` that will publish the job to NATS.

This feature requires a google data/youtube API key, which needs only be entered once. 
![image](https://github.com/user-attachments/assets/5c7b6fd5-87d1-43b6-9917-98d8f3a990ab)

The `Add Videos` button will not appear until the user has added the relevant API key. 


Then the video can be viewed upon clicking a valid link
![image](https://github.com/user-attachments/assets/a67bed57-ec11-4a33-b2f0-ebcfcda3cdbb)

calisio commented:
Does this have other changes besides the youtube additions? It seems like there are a bunch of files that have changes that are tied to linting maybe? Reviewing 75 files and over 2,000 lines of code seems tough

calisio commented:
<img width="1512" alt="Screenshot 2024-11-08 at 3 43 34 PM" src="https://github.com/user-attachments/assets/5c2198ac-3296-4754-8263-c05bea94514c">
This is the student view of videos- you probably shouldnt see "Video management" here, since the student wouldn't be managing the content. I don't even really think you really need that header at all honestly.

calisio commented:
<img width="1512" alt="Screenshot 2024-11-08 at 3 47 02 PM" src="https://github.com/user-attachments/assets/7a4da879-49cf-43f9-9550-7689b42b1a07">
In terms of real life usability, I think we should consider putting some sort of informational hover instructing the admin what a "youtube key" is, and how they would be able to get one. I only know what that is because you told me that it was in a developer account on google, but I had to look up how to get one for it to work. This can be a future ticket, but I definitely think it should be something to consider.
Something like this:

![image](https://github.com/user-attachments/assets/a1b7dbe3-1f0f-4ba8-a5dc-515278736629)

calisio commented:
After I put in the URL and it told me it was attempting to download, maybe we should refresh the page automatically in order to be able to view the video? In the future we should definitely include something in the UI that illustrates that we received the request for that specific video and display that the content is loading to keep the user up-to-date on the status of their request. If not, the user is left there waiting on an empty screen not knowing if it has gone through or not. In order to see if my video had gone through successfully, I had to refresh the page (where it then appeared).

On that note- if you refresh the page (which you have to) it sets the tab back to the default "Libraries" tab, even though we are on the video page. See the top tab view:
<img width="1512" alt="Screenshot 2024-11-08 at 3 55 54 PM" src="https://github.com/user-attachments/assets/0512f646-0695-45d8-a3dc-bd243486af97">

calisio commented on frontend/src/app.tsx:
I haven't looked too deep into this code, so if I'm completely off just ignore me. It seems like you condensed the admin view of open content management into the same page as the open content for the student. In developing the open content pages, I had left "Open Content" and "Open Content Management" as two separate pages because Chris and Nokie have been talking about being able to view student pages from the admin view (ie you would see what the student sees with open content, without the toggle buttons). I can still see that being a possibility (I think?) if we just change the role on that page to render the correct content, but let me know what you think. Again, I haven't reviewed the code in depth because there is a lot of it & I won't get into the details before the weekend. Just wanted to surface that in case it wasn't brought to your attention before.

PThorpe92 replied:
Solution: OpenContentManagement is it's own route, and I added a feature for an Admin to "View as Student", which will just direct the admin to the other route, and in the component: it will check whether it is an administrator viewing on that particular route: causing it to render itself as a student.

![image](https://github.com/user-attachments/assets/698b7f1c-1a52-41d9-9f57-9189fe46e5af)

![image](https://github.com/user-attachments/assets/74d48d2a-9b99-4316-999b-63a5c7ec8692)

### Libraries

![image](https://github.com/user-attachments/assets/2e182f4d-d762-493e-b292-c9fdfc420023)

![image](https://github.com/user-attachments/assets/204a5633-de94-494b-b439-f050ae24bdf9)

calisio commented:
As a user, if i put an invalid link (such as that of a YT profile), I should probably be notified that it is either an invalid link or that it has failed.
<img width="608" alt="Screenshot 2024-11-08 at 4 23 03 PM" src="https://github.com/user-attachments/assets/7b72750d-b0e2-467a-be76-a24a54d4f7c0">
<img width="1512" alt="Screenshot 2024-11-08 at 4 23 39 PM" src="https://github.com/user-attachments/assets/0fc80ab2-2ac0-4b4b-a612-abe5506ec1ab">
Right now, it shows nothing after the success toast (even after a refresh), since it does not create a card because it is not a video

calisio commented:
Should you be able to upload the same video twice?
<img width="606" alt="Screenshot 2024-11-08 at 4 31 18 PM" src="https://github.com/user-attachments/assets/de63f8a1-88bd-4b32-80b0-a5627935db96">

calisio commented:
Features I have tried that work:

- searching by the title
- adding multiple links WITH commas
- visible/hidden toggle & successfully changes it in the student portal
- viewing the video in the video viewer page

Features that do not work:

- Deleting a video does not work, it rather redirects you to the video viewing page. I assume that is because you have the onclick that redirects to the video viewing page to be on the entire card. You’ll need to increase the z-index of the trash can for it to trigger that button over the redirect.
- Sort does not work. If we dont need it at the moment, you can just take it out for now and we can add a ticket to add it into this page later.
- pagination stops at 10. Make sure to either add in pagination to the bottom of this page so you can move through the pages
- I would recommend returning the list of videos in a specific order (I think that for open content libraries it is sorted automatically by created_at), that way when a user updates the state from visible/hidden it always stays in the same spot. Right now, if you switch it to visible/hidden it jumps around to a different spot on the screen making it hard to keep track where it moved to.

calisio commented:
I can help clean up some of the other parts of the UI in this page so it stays consistent (sadly not before Monday). Overall this is a SICK feature and I cant wait for it to be shown off!! 😎

PThorpe92 commented:
>This is the student view of videos- you probably shouldnt see "Video management" here, since the student wouldn't be managing the content.
 - [x] fixed: 

> I think we should consider putting some sort of informational hover instructing the admin what a "youtube key" is

  - [x] I realized that I can just avoid using youtube's service for fetching the data and can instead rely in using yt-dlp itself, so we can do away with the notion of an api-key.

 >Should you be able to upload the same video twice?

  - [x] Good point, fixed

>(videos not showing up upon submission...) This makes me as the user assume that the YT link failed

 - [x]  Good idea. Added a polling system that will call `mutate()` every N seconds after videos are added, if no additional videos are fetched in N seconds, we wait N*2 seconds then request again until at least 1 of the videos requested shows up, or we reach a 10 second wait time. 
While this polling process continues, users are not allowed to manually request retries for obvious reasons. also changed the middleware to save + download each video in it's own goroutine so will be much faster.
![image](https://github.com/user-attachments/assets/dde42a37-9cc3-4810-bf52-f3919d116219)


>Sort 

  - [x] fixed

>Pagination 
 
  - [x] fixed

> Deletion

  - [x] fixed
>As a user, if i put an invalid link (such as that of a YT profile), I should probably be notified that it is either an invalid link or that it has failed.

- [x] :+1:  It's kinda difficult to tell just from a link whether it will actually work or not, because youtube has a few different formats that act as valid links. Also, since we can accept about any video site, we have to account for that and really can only give warnings that the link might be invalid. For this reason, I have added an ignore button so they have the ability to try it anyway.
![image](https://github.com/user-attachments/assets/0a8b2a3f-7540-497b-8117-7345d6c1c694)

> would recommend returning the list of videos in a specific order

- [x] defaults to created_at DESC unless they specify otherwise

PThorpe92 commented:
Added favorites 
![image](https://github.com/user-attachments/assets/a97f5254-074e-4907-9658-c156b710ebe5)

PThorpe92 commented:
Also added:

Videos will be hosted on s3, and many/all instances will share the same s3 bucket. Along with the videos stored as .mp4 files, we will also be storing `.json` metadata for each video, so if one instance downloads a video, it will be made available to all other instances to choose from, preventing any downloads from ever happening twice.

I added a cronjob that will fetch all .json from s3 to update the instances videos table so that they all stay up to date. We offer as many libraries as possible, so it would be ideal to offer as many videos as possible too and administrators can hide/show them as they please
